pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java

```diff
@@ -530,30 +574,58 @@ public KitRequestShipping getKitRequestShipping(@NonNull ResultSet rs) throws SQ
+    private Optional<ScanError> insertKitTrackingIfNotExists(KitRequestShipping kitRequestShipping, int userId) {
         Optional<ScanError> result = Optional.empty();
+        String errorMessage = String.format("Unable to insert tracking %s for %s.", kitRequestShipping.getTrackingId(),
+                kitRequestShipping.getKitLabel());
         SimpleResult results = inTransaction((conn) -> {
             SimpleResult dbVals = new SimpleResult();
+            try (PreparedStatement stmt = conn.prepareStatement(SELECT_KIT_TRACKING_BY_KIT_LABEL_OR_TRACKING_ID)) {
+                // If a kit exists with the given kit label or tracking id, the unique key will be violated
+                // on insert.  Return an error to the user with some information about the existing row.
+                stmt.setString(1, kitRequestShipping.getKitLabel());
+                stmt.setString(2, kitRequestShipping.getTrackingId());
+                try (ResultSet rs = stmt.executeQuery()) {
+                    if (rs.next()) {
+                        String kitLabel = rs.getString(DBConstants.KIT_LABEL);
+                        String trackingId = rs.getString(DBConstants.TRACKING_ID);
+                        String scannedAt = rs.getString(DBConstants.DSM_SCAN_DATE);
+                        String scannedBy = rs.getString(DBConstants.SCAN_BY);
+                        dbVals.resultValue = new ScanError(kitRequestShipping.getKitLabel(),
+                                String.format("Kit %s was already associated with tracking id %s by %s at %s",
+                                        kitLabel, trackingId, scannedBy, scannedAt));
+                        return dbVals;
+                    }
+                }
+            } catch (SQLException ex) {
+                dbVals.resultValue = new ScanError(kitRequestShipping.getKitLabel(), errorMessage);
+                logger.error(errorMessage, ex);
+            }
+
             try (PreparedStatement stmt = conn.prepareStatement(INSERT_KIT_TRACKING)) {
                 stmt.setLong(1, System.currentTimeMillis());
-                stmt.setString(2, userId);
+                stmt.setInt(2, userId);
                 stmt.setString(3, kitRequestShipping.getTrackingId());
                 stmt.setString(4, kitRequestShipping.getKitLabel());
+                stmt.setString(5, kitRequestShipping.getTrackingId());
+                stmt.setString(6, kitRequestShipping.getKitLabel());
                 int rowsAffected = stmt.executeUpdate();
                 if (rowsAffected != 1) {
-                    dbVals.resultValue = new ScanError(kitRequestShipping.getKitLabel(),
-                            String.format("Kit Label \"%s\" does not exist.\n"
-                                    + UserErrorMessages.IF_QUESTIONS_CONTACT_DEVELOPER, kitRequestShipping.getKitLabel()));
-                    logger.error(String.format("The number of affected rows for kit label %s is not 1", kitRequestShipping.getKitLabel()));
```

Comment:
The change here is that instead of erroring out when > 1 row is updated, we check for the row first and return a scan error without breaking the constraint and sending an alert.